src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,822 @@
+ * 2. "buffer_failbit" indicates an attempt to use the buffer when the buffer pointer
+ *    was nullptr (e.g. failure to allocate a unique_ptr of streambuf) (fileout/filein).
+ * 3. "string_failbit" indicates a failed operation in std::string related calls
+ *    (e.g. failure in pushing a line from the buffer to the given std::string in
+ *    get_line()) (filein only).
+ * 4. "posixio_errno" indicates that a POSIX I/O function failed, and the POSIX errno
+ *    was set (fileout/filein).
+ *
+ * If all of the above bits/variables are false or 0, the stream status is "good".
+ *
+ * The current_state() function returns the stream's current status based on this table.
+ */
+enum diostates
+{
+    eofbit = 0x01,
+    buffer_failbit = 0x02,
+    string_failbit = 0x04,
+    posixio_errno = 0x08
+};
+
+// Some empty classes for buffer operations implementation.
+// Used for accepting endl and flush in fileout::write() functions.
+class endline { };
+class flushbuf { };
+static constexpr endline endl;   // same as std::endl
+static constexpr flushbuf flush; // same as std::flush
+
+/* Specialise system_error & runtime_error for more meaningful exceptions on state bits/vars */
+// on eofbit
+class iostream_eof : public std::runtime_error
+{
+    public:
+    iostream_eof(const char *msg) : runtime_error(msg)
+    {
+    }
+};
+// on buffer_failbit, string_failbit
+class iostream_error : public std::runtime_error
+{
+    public:
+    iostream_error(const char *msg) : runtime_error(msg)
+    {
+    }
+};
+// on posixio_errno
+class posixio_error : public std::system_error
+{
+    public:
+    posixio_error(const int posix_errno) : system_error(posix_errno, std::system_category())
+    {
+    }
+};
+
+using streambuf = cpbuffer<IOSTREAM_BUFSIZE>;
+
+// fileinout_base is a base for fileout/filein which defines some general things.
+// Public functions are available in both fileout/filein and they does not get defined
+// at those classes.
+class fileinout_base
+{
+    protected:
+    int fd = -1;
+    std::unique_ptr<streambuf> buf;
+
+    public:
+    // Get raw pointer of current buffer
+    // NOTE: The buffer may be null if allocation failed (buffer_failbit will be set in this case)
+    streambuf *get_buf() noexcept;
+
+    // Is current stream's file descriptor open?
+    bool is_open() noexcept;
+
+    // Set file descriptor
+    // NOTE: Setting new fd will REPLACE current fd in the stream and It doesn't
+    // flush data to current fd before replacing it. Also It may result in leaked
+    // file descriptor (because stream use its fd to know what fd should be closed).
+    void set_fd(const int newfd) noexcept;
+
+    // Get current file descriptor
+    int get_fd() noexcept;
+
+    // Close current file descriptor if it is opened
+    bool close() noexcept;
+
+    // Return true if file is opened
+    operator bool() noexcept;
+
+    // Return true if file is closed
+    bool operator!() noexcept;
+};
+
+/*
+ * fileout
```

Comment:
should be `file_out`, please follow the style.

Reply:
Actually, could we call it "out_stream"? Or even "ostream"? It used to be called that but you changed it.

I think you misinterpreted an earlier comment where I pointed out you had copied the standard library iostream class hierararchy including the names. I did not mean that you needed to change the name.

Reply:
Reverted to `ostream`